### ISSUE TITLE
"word cancel" & so on don't work b/c of cancel command

### ISSUE THREAD

**Issue by rntz** [knausj85/knausj_talon#517]
The "cancel" command from https://github.com/knausj85/knausj_talon/blob/a953ecc55627fe1c07fc570d7b8cefd67dce1465/misc/cancel.talon is a bit too eager. It takes priority over basically any command ending in the word "cancel", eg. `word cancel` or `say I think I need to cancel`.

Options:

- Do nothing, you have to work around it by not using those commands, e.g. spelling the word cancel and avoiding phrases that end in it. I don't like this option, cancel is a common English word.

- Pick a rarer word. Also don't like this.

- Use two words. Maybe "cancel that" or "cancel cancel cancel". Then "word FOO" will never fail, and "phrase FOO" will only fail if you really need to say a phrase ending in "cancel that" (plausible) or "cancel cancel cancel" (unlikely).

- Petition @lunixbochs for a solution -- if 'cancel' could be a standalone command that automagically cancelled the whole phrase including prior commands, rather than hijacking <phrase> and anchoring, that might solve the issue nicely?

Anything I've missed?

**Comment by pokey** [comment]
This is a tough one. I think I'd lean towards 3) (multi-word command)

Not sure I follow 4) though. How would the behaviour change if it were baked into talon? Eg if you ended a phrase with cancel, would it not cancel? I personally would prob still want it to cancel in that situation. 

But if you did want the behaviour not to apply when it is after "word" or ends a phrase, you could prob have special cases for these, eg

```
word cancel: "cancel"
phrase <phrase> cancel$: insert(phrase)
```

Obv that last one is way oversimplified, but maybe could work? Prob simpler to just pick a longer command and let people override it if they want (your bullet 3)

**Comment by rntz** [comment]
I agree that a multi-word cancel is probably the best option for now. Any opinions on what it should be? I'd like to get this resolved sooner rather than later, before people get used to the current "cancel" command (now that it's in knausj by default).

I'm inclined to "cancel cancel"; it's unintuitive, but  reasonably short and not something that would normally occur in normal speech (unlike "cancel that" or "ignore that").

**Comment by knausj85** [comment]
cancel cancel sounds fine

**Comment by rntz** [comment]
The idea with adding a primitive magical cancel action or similar is that the command would then be `cancel` or `cancel$` instead of `^<phrase> cancel$`. This would make it get parsed more like most other commands, instead of (a) being able to match ANY utterance ending in `cancel` and (b) having high priority because of the start-anchoring.

(Also, it avoids having a command which starts with `<phrase>`, which could in principle make the decoder less accurate... This is sheer speculation, though; I should ask aegis for their opinion.)

**Comment by knausj85** [comment]
I think submitting an FR is a wise step too. I do think this command has enough merit as-is in in the interim, if it has no negative side effects.

**Comment by rntz** [comment]
Based on discussion with aegis in slack, this approach to a cancel command is a bad idea at least for now, because it is likely to significantly decrease recognition accuracy by screwing with the decoder's beam search. Some people seem to have already run into this. He suggests instead using a pre-phrase hook and (for instance) a `cancel cancel$` command. I think the idea is that the pre-phrase hook checks whether the command ended with "cancel cancel" (or whatever phrase we settle on) and if so cancels the phrase -- I forget what the interface for that is, though.
